rfcs/0000-rlp-v2.md

```diff
@@ -0,0 +1,740 @@
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore-simple-infra-rl
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - rates: # at least one rate limit required
+        - limit: 5
+          unit: second
+```
+
+##### Example 2. Specific route matching rule targeted, conditions, counter qualifiers and multiple rates
+
+In this example, a distinct limit is associated to each individual matching rule of the targeted HTTPRoute, using the `matches` field for fine-grained filtering. In both cases, the rate limits are to be enforced per username and only in case the user is not an admin.
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore-per-endpoint-per-user
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - name: readers
+      matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: GET
+      rates:
+        - limit: 50
+          duration: 1
+          unit: minute
+      counters:
+        - auth.identity.username
+      when:
+        - selector: auth.identity.group
+          operator: neq
+          value: admin
+
+    - name: writers
+      matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: POST
+      rates:
+        - limit: 5
+          duration: 1
+          unit: minute
+        - limit: 100
+          duration: 12
+          unit: hour
+      counters:
+        - auth.identity.username
+      when:
+        - selector: auth.identity.group
+          operator: neq
+          value: admin
+```
+
+##### Example 3. Route filtering using `when` conditions
+
+`when` conditions are preferably to be used for special cases of conditional filtering based on values other than attributes of the route that could otherwise be specified using the `matches` field. However, in some cases such as where the condition is not related to the HTTPRoute (e.g. see [Example 2](#example-2-specific-route-matching-rule-targeted-conditions-counter-qualifiers-and-multiple-rates) above; filterring based on hostname, etc) or a temporary measure while the target object misses the desired matching rule, the `when` conditions remains an option.
+
+In this example, a special limit with one rate limit of 150 rps is set for `GET` requests to the `/toys/special` path. Because the `toystore` HTTPRoute is unaltered, the special case is defined using the `when` conditions field.
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore-special-toys
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - name: read-special
+      matches:
```

Comment:
> Is the kuadrant control plane enforcing 1:1 match between policy `matches` and HTTPRoute matches?

It is.

> Because I feel this is assumed and nothing prevents having differences there. Only by a typo or mistake typing the policy there could not be a perfect match between RLP `matches` and HTTPRoute matches.

Correct. One typo and there will be a mismatch. This IMO the weakest point of this proposal. I would love to have a more assertive way to target individual HTTPRoute matches from a RLP.

> If the kuadrant control plane enforces 1:1 match between policy `matches` and HTTPRoute matches, what happens where there is no perfect match? The policy gets invalidated and reported in the status??

I would assume that invalidating the policy is likely the best thing to do. Reporting it in the status seems reasonable for cases where the mismatch originates in a change made to the HTTPRoute, while a validating webhook could help prevent such mismatches coming from the RLP itself.

Reply:
It'd be nice to have some way to track "inconsistencies" as they appear, e.g. when some `match` is changed within the `HTTPRouteRule` but the `RateLimitPolicy` isn't updated…

Reply:
https://github.com/Kuadrant/architecture/pull/8#discussion_r1123141323

Reply:
> It'd be nice to have some way to track "inconsistencies" as they appear,

Agreed. This is likely the weakest point of this proposal. I made sure to mention it in the [Drawbacks](https://github.com/Kuadrant/architecture/blob/rfc/rlp-cleanup/rfcs/0000-rlp-v2.md#drawbacks) section.